proposals/NNNN-atomics.md

```diff
@@ -0,0 +1,1702 @@
+
+```swift
+extension UnsafeAtomic where Value: AtomicInteger {
+  public func loadThenWrappingIncrement(
+    by delta: Value,
+    ordering: AtomicUpdateOrdering
+  ) -> Value
+  ...
+  public func bitwiseOrThenLoad(
+    with value: Value, 
+    ordering: AtomicUpdateOrdering
+  ) -> Value
+  ...
+  public func wrappingIncrement(
+    by delta: Value,
+    ordering: AtomicUpdateOrdering
+  )
+}
+
+let counter = UnsafeAtomic<Int>.create(initialValue: 0)
+defer { counter.destroy() }
+counter.wrappingIncrement(by: 42, ordering: .relaxed)
+```
+
+
+### Atomic Lazy References
+
+The operations provided by `UnsafeAtomic<Unmanaged<T>>` only operate on the unmanaged reference itself. They don't allow us directly access to the referenced object -- we need to manually invoke the methods `Unmanaged` provides for this purpose (usually, `takeUnretainedValue`).
+
+Note that loading the atomic unmanaged reference and converting it to a strong reference are two distinct operations that won't execute as a single atomic transaction. This can easily lead to race conditions when a thread releases an object while another is busy loading it:
+
+```swift
+// BROKEN CODE. DO NOT EMULATE IN PRODUCTION.
+let myAtomicRef = UnsafeAtomic<Unmanaged<Foo>>.create(initialValue: ...)
+
+// Thread A: Load the unmanaged value and then convert it to a regular
+//           strong reference.
+let ref = myAtomicRef.load(ordering: .acquiring).takeUnretainedValue()
+...
+
+// Thread B: Store a new reference in the atomic unmanaged value and 
+//           release the previous reference.
+let new = Unmanaged.passRetained(...)
+let old = myAtomicRef.exchange(new, ordering: .acquiringAndReleasing)
+old.release() // RACE CONDITION
+```
+
+If thread B happens to release the same object that thread A is in the process of loading, then thread A's `takeUnretainedValue` may attempt to retain a deallocated object.
+
+Such problems make `AtomicUnmanaged` exceedingly difficult to use in all but the simplest situations. The section on [*Atomic Strong References*](#atomic-strong-references-and-the-problem-of-memory-reclamation) below describes some new constructs we may introduce in future proposals to assist with this issue.
+
+For now, we provide the standalone type `UnsafeAtomicLazyReference`; this is an example of a useful construct that could be built on top of `UnsafeAtomic<Unmanaged>` operations. Of all the various atomic constructs introduced in this proposal, only `UnsafeAtomicLazyReference` represents a regular strong reference to a class instance -- the other pointer/reference types leave memory management entirely up to the user.)
+
+An `UnsafeAtomicLazyReference` holds an optional reference that is initially set to `nil`. The value can be set exactly once, but it can be read an arbitrary number of times. Attempts to change the value after the first `storeIfNilThenLoad` call are ignored, and return the current value instead.
```

Comment:
`storeIfNilThenLoad` ➡️ `storeIfNil` to match implementation?

```suggestion
An `UnsafeAtomicLazyReference` holds an optional reference that is initially set to `nil`. The value can be set exactly once, but it can be read an arbitrary number of times. Attempts to change the value after the first `storeIfNil` call are ignored, and return the current value instead.
```

Reply:
Ah, good eye -- I'm actually changing the implementation to match.